File: 4th Semester/Systems for Design and Implementation/C#/Homework4/Persistence/Database/RideDbRepository.cs
using log4net;
using log4net.Util;
using OnwardTravel.Models;
using OnwardTravel.Utils;
using System;
using System.Collections.Generic;
using System.Data;

namespace OnwardTravel.Repository.Database
{
    public class RideDbRepository
        : IRideRepository
    {
        private readonly DbUtils DbUtils;
        private static readonly ILog Logger = LogManager.GetLogger("RideDbRepository");

        public RideDbRepository(MyProperties props)
        {
            DbUtils = new DbUtils(props);
        }

        protected static Ride ExtractRide(IDataReader dataR)
        {
            Guid guid = dataR.GetGuid(0);
            string destination = dataR.GetString(1);
            DateTime departure = dataR.GetDateTime(2);
            return new Ride(guid, destination, departure);
        }

        public Guid? Save(Ride entity)
        {
            Logger.InfoFormat("Saving the ride: {0}", entity);
            using (var comm = DbUtils.GetConnection().CreateCommand())
            {
                comm.CommandText = "INSERT INTO Rides(destination, departure) VALUES (@destination, @departure)";
                var paramDestination = comm.CreateParameter();
                paramDestination.ParameterName = "@destination";
                paramDestination.Value = entity.Destination;
                comm.Parameters.Add(paramDestination);

                var paramDeparture = comm.CreateParameter();
                paramDeparture.ParameterName = "@departure";
                paramDeparture.Value = entity.Departure;
                comm.Parameters.Add(paramDeparture);

                int result = comm.ExecuteNonQuery();
                if (result == 0)
                {
                    Logger.InfoExt(null);
                    return null;
                }
            }

            using (var comm = DbUtils.GetConnection().CreateCommand())
            {
                comm.CommandText = "SELECT ride_id FROM Rides WHERE destination = @destination AND departure = @departure";
                var paramDestination = comm.CreateParameter();
                paramDestination.ParameterName = "@destination";
                paramDestination.Value = entity.Destination;
                comm.Parameters.Add(paramDestination);

                var paramDeparture = comm.CreateParameter();
                paramDeparture.ParameterName = "@departure";
                paramDeparture.Value = entity.Departure;
                comm.Parameters.Add(paramDeparture);

                var result = DbUtils.FindQueryResults(comm, DbUtils.ExtractGuid);
                if (result.Count == 0)
                {
                    Logger.InfoExt(null);
                    return null;
                }
                Logger.InfoExt(result[0]);
                return result[0];
            }
        }

        public Ride Find(Guid identifier)
        {
            throw new NotImplementedException();
        }

        public Ride Update(Ride entity)
        {
            throw new NotImplementedException();
        }

        public Ride Delete(Guid identifier)
        {
            throw new NotImplementedException();
        }

        public IList<Ride> FindAll()
        {
            using (var comm = DbUtils.GetConnection().CreateCommand())
            {
                comm.CommandText = "SELECT * FROM rides";
                return DbUtils.FindQueryResults(comm, ExtractRide);
            }
        }

        public IList<Ride> FindByDestinationAndDeparture(string destination, DateTime departure)
        {
            using (var comm = DbUtils.GetConnection().CreateCommand())
            {
                comm.CommandText = "SELECT * FROM rides WHERE destination LIKE CONCAT('%', @destination, '%') AND DATE_TRUNC('day', departure) = DATE_TRUNC('day', @departure)";
                var paramDestination = comm.CreateParameter();
                paramDestination.ParameterName = "@destination";
                paramDestination.Value = destination;
                comm.Parameters.Add(paramDestination);

                var paramDeparture = comm.CreateParameter();
                paramDeparture.ParameterName = "@departure";
                paramDeparture.Value = departure;
                comm.Parameters.Add(paramDeparture);

                return DbUtils.FindQueryResults(comm, ExtractRide);
            }
        }
    }
}

File: 4th Semester/Systems for Design and Implementation/Java/Homework2/src/main/java/OnwardTravel/repository/database/RideDbRepository.java
package OnwardTravel.repository.database;

import OnwardTravel.models.Ride;
import OnwardTravel.repository.RideRepository;
import OnwardTravel.repository.exceptions.DataLoadException;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Timestamp;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Properties;
import java.util.UUID;

public class RideDbRepository implements RideRepository {
    private final DbUtils dbUtils;
    private static final Logger logger = LogManager.getLogger(RideDbRepository.class);

    public RideDbRepository(Properties props) {
        logger.traceEntry("Initializing UserDbRepository with properties: {}", props);
        dbUtils = new DbUtils(props);
    }

    public static Ride extractRide(ResultSet resultSet) {
        try {
            UUID rideID = resultSet.getObject("ride_id", UUID.class);
            String destination = resultSet.getString("destination");
            LocalDateTime departure = resultSet.getTimestamp("departure").toLocalDateTime();
            return new Ride(rideID, destination, departure);
        } catch (SQLException ex) {
            logger.error(ex);
            throw new RuntimeException(ex);
        }
    }

    @Override
    public UUID save(Ride entity) throws DataLoadException {
        return null;
    }

    @Override
    public Ride find(UUID identifier) throws DataLoadException {
        return null;
    }

    @Override
    public Ride update(Ride entity) {
        return null;
    }

    @Override
    public Ride delete(UUID identifier) {
        return null;
    }

    @Override
    public List<Ride> findAll() throws DataLoadException {
        try (PreparedStatement preparedStatement = dbUtils.getConnection().prepareStatement(
                "SELECT * FROM rides"
        )) {
            return DbUtils.findQueryResult(preparedStatement, RideDbRepository::extractRide);
        } catch (SQLException ex) {
            logger.error(ex);
            throw new DataLoadException(ex);
        }
    }

    @Override
    public List<Ride> findByDestinationAndDeparture(String destination, LocalDateTime departure) throws DataLoadException {
        try (PreparedStatement preparedStatement = dbUtils.getConnection().prepareStatement(
                "SELECT * FROM rides WHERE destination LIKE CONCAT('%', ?, '%') AND departure = departure"
        )) {
            preparedStatement.setString(1, destination);
            preparedStatement.setTimestamp(2, Timestamp.valueOf(departure));
            return DbUtils.findQueryResult(preparedStatement, RideDbRepository::extractRide);
        } catch (SQLException ex) {
            logger.error(ex);
            throw new DataLoadException(ex);
        }
    }
}
